File: compiler/bench/catalan.fs
\ expected: 6564120420
\ Catalan number recursively, n=20

: catalan ( n -- c ) recursive
  dup 1 <= if drop 1 exit then
  0 over 0 do
    over 1- i - recurse
    i recurse
    * +
  loop nip ;

: main
  20 catalan . cr ;

File: compiler/bench/catalan.c
// expected: 6564120420
#include <stdio.h>
long catalan(long n) {
    if (n <= 1) return 1;
    long sum = 0;
    for (long i = 0; i < n; i++) {
        sum += catalan(n - 1 - i) * catalan(i);
    }
    return sum;
}

int main() {
    printf("%ld\n", catalan(20));
    return 0;
}
